docs/developer/using_trace.md

```diff
@@ -0,0 +1,262 @@
+`trace_entry` and `trace_exit` which should be used in all major functions. This
+makes it so the subprogram is logged in tracing.
+
+!!! note
+    The `trace_entry` and `trace_exit` routines do have an associated
+    cost, and therefore for short routines called very frequently (e.g. functions
+    in loops) it can be a bad idea for performance to trace these routines.
+
+The syntax for `trace_entry` is:
+
+```fortran
+call trace_entry(block_name, success[, profile][, class])
+```
+Where:
+
+- `block_name` - `character, intent(in)` - is the name of the entry in the
+  trace.
+- `success` - `integer, intent(out)` - is the return status of the function, `0` if
+  successful.
+- `profile` - `logical, intent(in)` - whether to profile this block
+- `class` - `character, intent(in)` - is the profile class for the purposes of
+  [`profclass`](#profclass-block) enabling profiling.
+
+The syntax for `trace_exit` is:
+```fortran
+call trace_exit(block_name, success)
+```
+
+Where:
+
+- `block_name` - `character, intent(in)` - is the name of the entry in the
+  trace.
+- `success` - `integer, intent(out)` - is the return status of the function, `0` if
+  successful.
+
+!!! note
+    `trace_entry` and `trace_exit` subroutines can be called around any
+    arbitrary blocks of code. This makes it possible to narrow down failing
+    points, but not only that, they can be used to profile sections of code
+    where performance is falling below expectations.
+
+#### Trace Sections
+
+It is possible to mark a section of a code as a sub-element of a trace-block
+rather than a block in its own right. These can also have extra arguments passed
+through to log parameters at certain stages.
+
+The syntax for `trace_section` is:
+
+```fortran
+call trace_section_entry(section_name[, variable_name, value], success[, class])
+```
+Where:
+
+- `section_name` - `character, intent(in)` - is the name of the entry in the
+  trace.
+- `variable_name` - `character, intent(in)` - is the logging name to be attached
+  to the section in the trace output
+- `value` - `integer|character, intent(in)` - is the current value of the
+  variable to be logged. `value` can either be an `integer` or
+  `character(len=63)`
+- `success` - `integer, intent(out)` - is the return status of the function, `0` if
+  successful.
+- `class` - `character, intent(in)` - is the profile class for the purposes of
+  `profclass` enabling profiling.
+
+The syntax for `trace_exit` is:
+
+```fortran
+call trace_section_exit(block_name, success)
+```
+
+Where:
+
+- `block_name` - `character, intent(in)` - is the name of the entry in the
+  trace.
+- `success` - `integer, intent(out)` - is the return status of the function, `0` if
+  successful.
+
+### Debugging
+
+When castep performs an `io_abort`, it first calls into `trace` on each process
+to dump the current trace stack to the `<seedname>.<process_id>.err` files. This
+is the main error message people see when castep fails.
```

Comment:
I suggest this is moved further up, as it's an obvious benefit even if you aren't doing specific debugging or profiling.